qpdf/11.1.1 add (lib only) package

cguentherTUChemnitz opened the pull request:
Specify library name and version:  **qpdf/11.1.0**

I am not a maintainer of the qpdf project, but i would like to have it in the conan package ecosystem. The project speaks best on its-self:
https://github.com/qpdf/qpdf

The project provides cli tools as well as c++ and c libraries.

One explanation for the decision to go for **libjpeg-turbo as hardcoded requirement**:
libjpeg produces runtime errors in the package-test with:
```
error while getting stream data for 7 0 R: Wrong JPEG library version: library is 90, caller expects 80
```

This seems that i would have to downgrade libjpeg to something that is not present on the conan repository. Nevertheless i was happy to find out that libjpeg-turbo was just working fine.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=13113) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=13113) before we can accept your contribution.<br/><hr/>**Christian Günther** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=13113) it.</sub>

cguentherTUChemnitz commented:
Mh it seems there is an requirement from within qpdf to require a more modern cmake version that installed on the ci-runners, since they report:

```
-- Configuring incomplete, errors occurred!
qpdf/11.1.0: 
CMake Error at CMakeLists.txt:2 (cmake_minimum_required):
  CMake 3.16 or higher is required.  You are running version 3.15.7
```

What is the way here to go? I an upgrade on ci-systems necessary? That would wonder me since other packages are using cmake version 3.16 in there test_package, like in boost-ext-ut/all/test_package/CMakeLists.txt. So how such a package was able to merge? Would it be a solution to add cmake from conan in a more modern version as a build_requirement?

uilianries commented:
@cguentherTUChemnitz You should add cmake as tool_requires, in build_requirements() method: https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#requirements

uilianries reviewed (commented):
Thank you for your contribution. Please, refine the template according to your target project, remove unused methods and comments.

cguentherTUChemnitz commented:
> Thank you for your contribution. Please, refine the template according to your target project, remove unused methods and comments.

Thanks for your pleasant examples and documentations as well as your fast and detailed feedback ;) I will do it shortly

cguentherTUChemnitz commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,138 @@
+        tc.variables["REQUIRE_CRYPTO_OPENSSL"] = False
+        if is_msvc(self):
+            # don't use self.settings.compiler.runtime
+            tc.cache_variables["USE_MSVC_RUNTIME_LIBRARY_DLL"] = not is_msvc_static_runtime(self)
```
That was just part of the template. I don't have an MSVC at my hand, so i can't confirm if this is necessary. When you doubt it yourself, then i will just remove it as long as no test and no bugreport will state otherwise ;)

cguentherTUChemnitz commented:
@uilianries I tried to update everything as your review suggest.
 - removed a lot of comments from the templates
 - shrinked down the test to some version printing instead of doing real work
 - introduced with_crypto option to select between native and openssl (didn't find the third way of gnutls on conan)
 - revisited the linked system libs (only libm seems to be necessary here, as mentioned from the great conan ci-builds)
 
 Do you think this is now in a mergeable state?

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,131 @@
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "with_crypto": ["native","openssl"],
```
```suggestion
        "with_crypto": ["native", "openssl", "gnutls"],
```
Let's add the option now, and keep it as restricted, so as soon as we have that package we just enabled it. The advantage: It should not affect the package Id in the future.

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,131 @@
+    def requirements(self):
+        # https://qpdf.readthedocs.io/en/stable/installation.html#basic-dependencies
+        if self.options.with_crypto == "openssl":
+            self.requires("openssl/3.0.5")
```
```suggestion
            self.requires("openssl/3.0.5")
        elif self.options.with_crypto == "gnutls":
            raise ConanInvalidConfiguration("GnuTLS is not available in Conan Center yet.")
```

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,131 @@
+
+
+    def validate(self):
+        # validate the minimum cpp standard supported. For C++ projects only
```
```suggestion
```
Keep only comments related to the project. For recipes, it's all documented already.

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,131 @@
+            tc.variables["REQUIRE_CRYPTO_OPENSSL"] = True
+
+        tc.generate()
+        tc = CMakeDeps(self) # init requirements from dependencies
```
```suggestion
        tc = CMakeDeps(self)
```

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,131 @@
+        cmake = CMake(self)
+        cmake.install()
+
+        # some files extensions and folders are not allowed. Please, read the FAQs to get informed.
```
```suggestion
```

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,131 @@
+        cmake.install()
+
+        # some files extensions and folders are not allowed. Please, read the FAQs to get informed.
+        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
```
Are you sure that project is generating all those files? .la, .pdb, .pc? Please, remove what's not used, check the build log after passing by the CI.

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,131 @@
+        self.cpp_info.set_property("cmake_target_name", "qpdf::libqpdf")
+        self.cpp_info.set_property("pkg_config_name", "libqpdf")
+
+        # If they are needed on Linux, m, pthread and dl are usually needed on FreeBSD too
```
```suggestion
```

uilianries commented on recipes/qpdf/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+find_package(QPDF REQUIRED CONFIG)
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+# don't link to ${CONAN_LIBS} or CONAN_PKG::package
```
```suggestion
```

uilianries commented on recipes/qpdf/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.8)
+
+project(test_package CXX) # if the project uses c++
```
```suggestion
project(test_package CXX)
```

uilianries commented on recipes/qpdf/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+add_executable(${PROJECT_NAME} test_package.cpp)
+# don't link to ${CONAN_LIBS} or CONAN_PKG::package
+target_link_libraries(${PROJECT_NAME} PRIVATE qpdf::libqpdf)
+# In case the target project need a specific C++ standard
```
```suggestion
```

uilianries commented on recipes/qpdf/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,30 @@
+import os
+
+
+# It will become the standard on Conan 2.x
```
```suggestion
```
Again, template comments should be removed.

uilianries commented on recipes/qpdf/all/test_package/test_package.cpp:
```diff
@@ -0,0 +1,9 @@
+int
+main(int argc, char* argv[])
+{
+    std::cout << "QPDF_VERSION " << QPDF_VERSION << "\n";
```
It's not enough. QPDF_VERSION is a define so we are consuming only a header, but not testing the library linkage. Please, add a function from the library to this code.

cguentherTUChemnitz replied:
Ok yeah, also good point. I added now an as minimal as possible write-out of an empty pdf to ensure that there is run-time dependency of linked functionality that can't be optimized away. I hope the new test_package content is suitable now

cguentherTUChemnitz commented:
@uilianries fine that way?

`continuous-integration/jenkins/pr-merge Expected — Waiting for status to be reported`
Do i have to do something here, or is it just wait for the scheduled build job?

cguentherTUChemnitz commented:
@uilianries  Are we finished here? Can we trigger additional people to provide the necessary reviews?

uilianries reviewed (dismissed):
LGTM

uilianries commented:
@cguentherTUChemnitz please, add a test_v1_package so we can validate with Conan v1 too. There is a good template here: https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package/all/test_v1_package

cguentherTUChemnitz commented:
> @cguentherTUChemnitz please, add a test_v1_package so we can validate with Conan v1 too. There is a good template here: https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package/all/test_v1_package

Do you know how i can test it locally? When i run:
```
conan create . qpdf/11.1.0@ -pr:b=default -pr:h=default --build=missing
```
I am not aware if it is running test_v1_package at all. It seems for me that this is not part of the execution for now.

uilianries commented:
@cguentherTUChemnitz Almost there, you missed `-tf`:

    conan create . qpdf/11.1.0@ -pr:b=default -pr:h=default --build=missing -tf test_v1_package
    
Also, you can skip build part and only execute the test:

    conan test test_v1_package qpdf/11.1.0@ -pr:b=default -pr:h=default

cguentherTUChemnitz commented:
@uilianries @ericLemanissier  can we trigger here the last missing reviews? Would be great when we can merge it soon ;)

cguentherTUChemnitz commented:
@uilianries @danimtb @SpaceIm @SSE4 
Hi there, just gathered here some people i did see more often in this repo. Can we please provide here the necessary review approvals? I think this PR is already in a good state, thanks to a lot of review help of uilianries. It would be nice, when we can merge this quite soon.

jwillikers commented on recipes/qpdf/all/test_v1_package/test_package.cpp:
```diff
@@ -0,0 +1,27 @@
+#include <qpdf/DLL.h>
```
I think you meant to delete this file.

cguentherTUChemnitz replied:
uilianries requested to add the test_v1_package scenario to run also the "old" conan package world. I do not like the code-duplication here for the source file "test_package.cpp". What do you think would be the best solution here?

cguentherTUChemnitz replied:
ahh now i understand. Thanks for pointing out. I remove the unnecessary file here

jwillikers reviewed (changes requested):
Thanks for the contribution! Looks very good.

cguentherTUChemnitz commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,134 @@
+    def _compilers_minimum_version(self):
+        return {
+            "gcc": "5",
+            "Visual Studio": "19.10",
```
I think it would be beneficial to update the cmake package template accordingly, since i took it from there and just updated the version which lists C++14 support, as demanded from qpdf.

I am still quite confused what to do here exactly. Should i add another entry for msvc, or should i replace the "Visual Studio" entry? I don't get the desired usage from just the function documentation.

jwillikers replied:
Wow! I didn't know those package templates were there. Those are incredibly helpful! I'll see if I can do a PR to update that template later / get feedback on the best way to do that. I use `check_min_vs()` in the `boost-ext-ut` recipe: https://github.com/conan-io/conan-center-index/blob/master/recipes/boost-ext-ut/all/conanfile.py#L59-L68

jwillikers commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,134 @@
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["BUILD_SHARED_LIBS"] = self.options.shared
```
Yes, an `option()` sets it a CMake cache variable. You should be able to omit this line then.

cguentherTUChemnitz replied:
qpdf seems to handle shared and static lib option independently from each other. Are both variables `BUILD_SHARED_LIBS` and `BUILD_STATIC_LIBS` are handled by your mentioned automatic management of conan?

jwillikers replied:
`BUILD_STATIC_LIBS` isn't a standard CMake variable and it isn't set by Conan, so you should leave the line for configuring that variable.

jwillikers commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -4,6 +4,7 @@
 from conan.tools.build import check_min_cppstd
 from conan.tools.scm import Version
 from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.microsoft.visual import is_msvc, check_min_vs
```
You should be able to drop the `.visual` here.

cguentherTUChemnitz commented on recipes/qpdf/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,28 @@
+    def test(self):
+        if can_run(self):
+            self.run(os.path.join(self.cpp.build.bindirs[0], "test_package") + " out_empty.pdf", env="conanrun")
+            self.run(os.path.join(self.deps_cpp_info["qpdf"].rootpath, "bin", "qpdf") + " --version", env="conanrun")
```
@jwillikers can you check if my current changes are reflecting it correctly? I just removed the manual path selection and it seemed to work out of the box to find correctly the binaries created by this package. Is this than already sufficient?

jwillikers replied:
I wouldn't expect `qpdf` to be available on the `PATH` without having to specify anything. @uilianries Does that seem right?

jwillikers commented:
> @uilianries @jwillikers Does someone have an idea how to cope with this problem? I think i need help here, since i don't have any windows machine on my hand.

So, `qpdf` uses PkgConfig and not native CMake find/config modules. That means you need to use the [PkgConfigDeps](https://docs.conan.io/en/latest/reference/conanfile/tools/gnu/pkgconfigdeps.html) generator. You may not even need the `CMakeDeps` generator. You can see the call to [pkg_check_modules](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html#command:pkg_check_modules) used by `qpdf` to find `libjpeg` below.

```sh
pkg_check_modules(pc_libjpeg libjpeg)
```

I'm not familiar with `qpdf` so not exactly sure why it may be failing only on Windows and not elsewhere.

Also, I believe you want to include the [VirtualBuildEnv](https://docs.conan.io/en/latest/reference/conanfile/tools/env/virtualbuildenv.html) generator in your recipe to ensure the `build_requirements` are made available to the recipe at build time.

cguentherTUChemnitz commented:
> CMakeDeps

Thanks for pointing out the direction here. It sound reasonable for the error. Nevertheless, i am not able to come around the not finding warnings in my local run in the conanio/gcc11-ubuntu16.04 container.  I hope at least it is fixing the windows side at the moment, but it leaves a bad gut-feeling that the linux build is still missing out those libs, even be able to compile nevertheless.

uilianries commented:
@cguentherTUChemnitz @jwillikers It's related to https://github.com/conan-io/conan/issues/11962

CMakeToolchain does not configure PKG_CONFIG_PATH properly, we need to set it manually.

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,143 @@
+        tc = CMakeDeps(self)
+        tc.generate()
+        tc = VirtualBuildEnv(self)
+        tc.generate(scope="build")
```
```suggestion
        tc.generate(scope="build")
        env = Environment()
        env.prepend_path("PKG_CONFIG_PATH", self.generators_folder)
        envvars = env.vars(self)
        envvars.save_script("pkg_config")
```
Let's set PKG_CONFI_PATH manually.

uilianries commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,143 @@
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.microsoft import is_msvc, check_min_vs
+from conan.tools.gnu import PkgConfigDeps
+from conan.tools.env import VirtualBuildEnv
```
```suggestion
from conan.tools.env import VirtualBuildEnv, Environment
```

uilianries commented:
I tested locally those suggestions and it works partially:

```
-- Performing Test WMAIN_WORKED1 - Success                                                                                                                     
-- Found PkgConfig: C:/Users/uilia/.conan/data/pkgconf/1.9.3/_/_/package/0a420ff5c47119e668867cdb51baff0eca1fdb68/bin/pkgconf.exe (found version "1.9.3")      
-- Checking for module 'zlib'                                                                                                                                  
--   Found zlib, version 1.2.12                                                                                                                                
-- Checking for module 'libjpeg'                                                                                                                               
--   Found libjpeg, version 9e                                                                                                                                 
--                                                                                                                                                             
-- *** Crypto Summary ***                                                                                                                                      
--   GNU TLS crypto enabled: OFF                                                                                                                               
--   OpenSSL crypto enabled: OFF                                                                                                                               
--   Native crypto enabled: ON                                                                                                                                 
--   Default crypto: native                                                                                                                                    
--                                                                                                                                                             
-- Looking for sys/types.h                                                                                                                                     
-- Looking for sys/types.h - found                                                                                                                             
```

It found zlib and libjpeg ...

```
﻿﻿﻿C:\Users\uilia\.conan\data\qpdf\11.1.1\_\_\build\bd28c61d36fa351a67f6c0abba4b742cd27de759\src\include\qpdf/Pl_DCT.hh(32,10): fatal error C1083: Cannot open include file: 'jpeglib.h': No such file or directory (compiling source file C:\Users\uilia\.conan\data\qpdf\11.1.1\_\_\build\bd28c61d36fa351a67f6c0abba4b742cd27de759\src\libqpdf\Pl_DCT.cc) [C:\Users\uilia\.conan\data\qpdf\11.1.1\_\_\build\bd28c61d36fa351a67f6c0abba4b742cd27de759\build\libqpdf\libqpdf_object.vcxproj]
  Pl_LZWDecoder.cc
  Pl_MD5.cc
  Pl_OStream.cc
  Pl_PNGFilter.cc
  Pl_QPDFTokenizer.cc
  Pl_RC4.cc
C:\Users\uilia\.conan\data\qpdf\11.1.1\_\_\build\bd28c61d36fa351a67f6c0abba4b742cd27de759\src\libqpdf\Pl_Flate.cc(5,10): fatal error C1083: Cannot open include file: 'zlib.h': No such file or directory [C:\Users\uilia\.conan\data\qpdf\11 .1.1\_\_\build\bd28c61d36fa351a67f6c0abba4b742cd27de759\build\libqpdf\libqpdf_object.vcxproj]
```

But failed to find headers ...

uilianries commented:
It will require a patch to inject cmake targets for qpdf ...

cguentherTUChemnitz commented:
> It will require a patch to inject cmake targets for qpdf ...

@uilianries  Do you have maybe already an idea what type of patching needs to be done?

cguentherTUChemnitz commented:
I think i solved the situation at least for libqpdf (with the necessity to remove qpdf binaries from compilation targets by cmake patches). I pushed a fix for the ci before the ci finished. So it seemes that this MR did get a Failed flag of the ci based on a failed build, which is already fixed by the last commit. 
@uilianries @jwillikers Is this possible to retrigger in some kind to get a green pipeline here accordingly to my last commit?

uilianries commented:
@cguentherTUChemnitz if you need to re-trigger a build, you can close the PR, wait 10 seconds, then re-open it again. It will trigger the CI again to re-build failed jobs. To re-build succeed job, you need to commit something, it can be an empty commit.

uilianries commented:
>
>         "_jpeg_std_error", referenced from:
>             Pl_DCT::finish() in Pl_DCT.cc.o
>         "_jpeg_write_scanlines", referenced from:
>             Pl_DCT::compress(void*, Buffer*) in Pl_DCT.cc.o
>       ld: symbol(s) not found for architecture x86_64
>       clang: error: linker command failed with exit code 1 (use -v to see invocation)

Weird error 🤔

cguentherTUChemnitz commented:
> > ```
> >     "_jpeg_std_error", referenced from:
> >         Pl_DCT::finish() in Pl_DCT.cc.o
> >     "_jpeg_write_scanlines", referenced from:
> >         Pl_DCT::compress(void*, Buffer*) in Pl_DCT.cc.o
> >   ld: symbol(s) not found for architecture x86_64
> >   clang: error: linker command failed with exit code 1 (use -v to see invocation)
> > ```
> 
> Weird error thinking

That is really strange. the compile log for the other mac ci jobs are highlighting that this is only the case for shared builds. The statics builds are able to compile, link and test. Is there some possibility to replay this locally? I have potentially also access to a mac environment. The most interesting part here in my eyes is, if the symbols are present in the object file at all and if really only the architecture as a mismatch here.

cguentherTUChemnitz commented:
@uilianries is it possible in your eyes to black-list at the moment this one combination of mac-platform with shared lib option, as known broken, that we can at least merge the others? I really have some time-pressure to get this lib in the official build channels t least for the linux build paths.

jwillikers commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,165 @@
+        tc = CMakeToolchain(self)
+        tc.variables["BUILD_STATIC_LIBS"] = not self.options.shared
+        # https://qpdf.readthedocs.io/en/latest/installation.html#build-time-crypto-selection
+        tc.variables["USE_IMPLICIT_CRYPTO"] = False
```
Try defining `tc.cache_variables["CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP"] = True` here to fix your Windows error.

```suggestion
tc.variables["USE_IMPLICIT_CRYPTO"] = False
tc.cache_variables["CMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP"] = True
```

cguentherTUChemnitz replied:
I added in hope to fix. It seems that the problem may be the same for mac. When this come green through, i will try to re-enable the mac shared builds once more.

jwillikers replied:
This will fix an issue specific to your Windows builds failing. I saw your fix for macOS shared libraries. Since they are using an `OBJECT` library, declaring `PRIVATE` linkage won't actually propagate to wherever the library is actually linked. So, your fix looks correct there. I hope it works.

jwillikers commented on recipes/qpdf/all/patches/0001-libqpdf-cmake-deps-jpeg-zlib.patch:
```diff
@@ -0,0 +1,25 @@
+     PUBLIC
+     $<BUILD_INTERFACE:${qpdf_INCLUDE}>
+     $<INSTALL_INTERFACE:include>)
+-  target_link_directories(${SHARED_LIB} PRIVATE ${dep_link_directories})
+-  target_link_libraries(${SHARED_LIB} PRIVATE ${dep_link_libraries})
```
You will also need to be sure to use `target_link_libraries(${SHARED_LIB} PUBLIC JPEG::JPEG ZLIB::ZLIB)` here and likewise for `STATIC_LIB` further down. Note that the shared / static libraries may not be built from the corresponding `OBJECT_LIB`, in which case, the dependencies won't be propagated to the target as they should be.

jwillikers replied:
The code here is the issue:

```
  if(OBJECT_LIB_IS_PIC)
    add_library(${SHARED_LIB} SHARED $<TARGET_OBJECTS:libqpdf_object>)
  else()
    add_library(${SHARED_LIB} SHARED ${libqpdf_SOURCES})
  endif()
```

`OBJECT_LIB_IS_PIC` will be false because `BUILD_STATIC_LIBS` is set to false when building shared libraries. Therefore, the shared library is not built from the object library and does not receive the link libraries from the `target_link_libraries` calls on the `OBJECT_LIB`.

cguentherTUChemnitz replied:
Yeah you are right. I hoped the object lib will transport it automatically as a public dependency. Nevertheless, to make sure it is really consumed by the other targets, i will now just hard-code add it to them also.

jwillikers commented on recipes/qpdf/all/patches/0001-libqpdf-cmake-deps-jpeg-zlib.patch:
```diff
@@ -0,0 +1,50 @@
+-  target_link_directories(${SHARED_LIB} PRIVATE ${dep_link_directories})
+-  target_link_libraries(${SHARED_LIB} PRIVATE ${dep_link_libraries})
+   if(ATOMIC_LIBRARY)
+-    target_link_libraries(${SHARED_LIB} PRIVATE ${ATOMIC_LIBRARY})
```
Do you need to leave this line unchanged and put the `target_link_libraries` for JPEG / ZLIB outside the `if(ATOMIC_LIBRARY)` check?

jwillikers replied:
I just confirmed that `ATOMIC_LIBRARY` will be false on macOS. From the console output in the macOS logs, you can see that `ATOMIC_WORKED0` is a success:

```
-- Performing Test ATOMIC_WORKED0
-- Performing Test ATOMIC_WORKED0 - Success
``

`qpdf` only sets `ATOMIC_LIBRARY` when `ATOMIC_WORKED0` fails and `ATOMIC_WORKED1` succeeds.
Since it isn't set, JPEG and ZLIB are not linked in.

cguentherTUChemnitz replied:
Oh good point. I hope my new patch-set is more versatile, since i set now the dependencies in an additional target_link_libraries call outside of the conditional checks for every library type (object, static, shared)

jwillikers commented:
Looks like you just need to patch out the calls to `find_library` and `find_path` for ZLIB and JPEG.

Just remove this entire section in your patches: https://github.com/qpdf/qpdf/blob/11.1/libqpdf/CMakeLists.txt#L152-L190

cguentherTUChemnitz commented:
> find_library

It seems that this is not enough when openssl comes into play. Then i get the same offor of not infding -lssl und -lcrypto. Locally in the conan linux gcc container i get a working state, when i just avoid the two calls from my last commit. This should disable the external dependency linkage from qpdf completely, only using my patch-in ones.

cguentherTUChemnitz commented:
It seems that we finally got a step forward, since mac is enabled and running the tests on static and dynamic linking.

Nevertheless, we see all windows tests failing with very curious erros during compilation:

```
----Running------
> cmake --build C:\J\w\prod\BuildSingleReference@2\conan-center-index\recipes\qpdf\all\test_v1_package\build\a1f410b7f918b45b1ec298bbba1d330c7bc32c54 --config Release -- /m:8 /verbosity:minimal
-----------------
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/J/w/prod/BuildSingleReference@2/conan-center-index/recipes/qpdf/all/test_v1_package/CMakeLists.txt
  test_package.cpp
C:\J\w\prod\BuildSingleReference@2\.conan\data\qpdf\11.1.1\_\_\package\26bafdc4ba785a8a3a974d957f1b04f467f69663\include\qpdf/PointerHolder.hh(31): fatal error C1021: invalid preprocessor command 'warning' [C:\J\w\prod\BuildSingleReference@2\conan-center-index\recipes\qpdf\all\test_v1_package\build\a1f410b7f918b45b1ec298bbba1d330c7bc32c54\test_package.vcxproj]
´``

Since i lack the access of a windows machine and msvc compiler, i would disable the platform here from my side. This can be enabled by someone else with msvc access afterwards and should hopefully not block the current development. I have found this documentation:
https://github.com/qpdf/qpdf/blob/main/README-windows.md
but i am totally not aware here, if some magic dependency stuff between msys2 and msvc is necessary here.

jberkenbilt commented:
> With this insight, i would open a ticket on qpdf repository, and disable msvc as compiler here for now, as long as the new ticket is unsolved.

(qpdf maintainer here) You can make sure that the `POINTERHOLDER_TRANSITION` preprocessor symbol is defined as `4` for the build of libqpdf, though you should allow users to define whatever value is appropriate for them. New users should use `4`. See https://qpdf.readthedocs.io/en/stable/design.html#smart-pointers for a full discussion.

jwillikers commented on recipes/qpdf/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_14)
+# msvc has problems consuming #warning macro
+# therefore we need a code-path in the include avoiding this warning https://github.com/qpdf/qpdf/issues/804
+target_compile_definitions(${PROJECT_NAME} PUBLIC POINTERHOLDER_TRANSITION=4)
```
You also need to add this to the `test_v1_package`.

cguentherTUChemnitz replied:
jeah sure, i have overseen this

cguentherTUChemnitz commented:
@uilianries @jwillikers Jeah finally working states for all major platforms :) Now we can gather the missing reviews ;)

jwillikers commented:
Awesome job! Thanks for sticking with it. FYI, it's usually alright to add a package with support for a limited set of platforms and note that "Contributions are welcome" for adding support for the others. I figured we could push through it and get this working. And you did!

cguentherTUChemnitz commented:
> Awesome job! Thanks for sticking with it. FYI, it's usually alright to add a package with support for a limited set of platforms and note that "Contributions are welcome" for adding support for the others. I figured we could push through it and get this working. And you did!

Thanks also for all your fast insights and reviews. I really like the pace and quality of this project here ;)

jwillikers commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,164 @@
+            self.cpp_info.system_libs.append("m")
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = "QPDF"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "QPDF"
+        self.cpp_info.names["cmake_find_package"] = "QPDF"
+        self.cpp_info.names["cmake_find_package_multi"] = "QPDF"
```
```suggestion
        self.cpp_info.filenames["cmake_find_package"] = "qpdf"
        self.cpp_info.filenames["cmake_find_package_multi"] = "qpdf"
        self.cpp_info.names["cmake_find_package"] = "qpdf"
        self.cpp_info.names["cmake_find_package_multi"] = "qpdf"
        self.cpp_info.components["libqpdf"].names["cmake_find_package"] = "libqpdf"
        self.cpp_info.components["libqpdf"].names["cmake_find_package_multi"] = "libqpdf"
```

jwillikers commented on recipes/qpdf/all/test_v1_package/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+find_package(QPDF REQUIRED CONFIG)
+
+add_executable(${PROJECT_NAME} ../test_package/test_package.cpp)
+target_link_libraries(${PROJECT_NAME} PRIVATE QPDF::QPDF)
```
```suggestion
target_link_libraries(${PROJECT_NAME} PRIVATE qpdf::libqpdf)
```

jwillikers reviewed (changes requested):
Re-combing the changes, I found a few more things that should probably be tweaked. Most notably, `qpdf`'s provided CMake Config module uses slightly different naming than the one here. That needs to match what the upstream provides.

cguentherTUChemnitz commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,164 @@
+            tc.variables["REQUIRE_CRYPTO_GNUTLS"] = False
+            tc.variables["REQUIRE_CRYPTO_OPENSSL"] = True
+        tc.generate()
+        tc = PkgConfigDeps(self)
+        tc.generate()
```
Good question. I think we would only loose the print-out during detection. But those detected are not anymore used after my cmake patches. I don't know if it is better to have the found-printouts, that later-on the missing dependency prints are not confusing for other looking in the build-log. What would you prefer?

jwillikers replied:
I think the missing print-out will be easier to track. Someone looking at the errors is likely to think the dependencies are coming from `pkg-config` instead of CMake modules, which could lead to a lot of wasted time when debugging. Using `find_package` in your patches should provide at least some output to the console, too.

cguentherTUChemnitz replied:
Ok i see your point and i go with it.

I can confirm on local builds, that it is not necessary anymore. So i will remove it. Nevertheless, i try to add an explanatory comment on why we don't go for pkgconfig, even when the qpdf uses that by default

jwillikers commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,164 @@
+        rmdir(self, os.path.join(self.package_folder, "share"))
+
+    def package_info(self):
+        self.cpp_info.libs = ["qpdf"]
```
Should this be `libqpdf`?

```suggestion
        self.cpp_info.libs = ["libqpdf"]
```

cguentherTUChemnitz replied:
as soon as i add this "lib" before it, it fails locally with:

```
CMake Error at build/generators/cmakedeps_macros.cmake:39 (message):
  Library 'libqpdf' not found in package.  If 'libqpdf' is a system library,
  declare it with 'cpp_info.system_libs' property
Call Stack (most recent call first):
  build/generators/qpdf-Target-release.cmake:34 (conan_package_library_targets)
  build/generators/qpdfTargets.cmake:26 (include)
  build/generators/qpdf-config.cmake:10 (include)
  CMakeLists.txt:5 (find_package)
```

I encountered previously that cmake does add in some scenarios the "lib"-keyword automatically. I don't know if this is here part of the magic that happens. Nevertheless it seems to correctly map with "qpdf" to libqpdf.a file on the disk.

jwillikers replied:
Nevermind, my bad. The install command changes the name to `qpdf`.

cguentherTUChemnitz replied:
Mh it seems you are right, that the package definition is not correctly, since some plattform tests still fails. 

When i reapply 
```
        self.cpp_info.components["libqpdf"].names["cmake_find_package"] = "libqpdf"
        self.cpp_info.components["libqpdf"].names["cmake_find_package_multi"] = "libqpdf"
```
and use for cpp_info.libs = "libqpdf", than the cmake found error dissolves, but the linker does not resolve correctly the symbols anymore again.

I am here quite out of knowledge how the CMakeLists configuration should be designed via conanfile, to match the official one. @jwillikers can you provide a tested setup for the package_info setup, that is in sync with the qpdf upstream cmake file? Otherwise i would like to just revert the the working state before that is not in sync with qpdf upstream, but was able to get the ci-green on all plattforms.

jwillikers replied:
Yep, the components have to define relationships to dependencies explicitly. Oh and, ` self.cpp_info.libs = ["qpdf"]` is correct. The `qpdf` project renames the library's output name to `qpdf` which I'd missed.

cguentherTUChemnitz commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,164 @@
+        env = Environment()
+        env.prepend_path("PKG_CONFIG_PATH", self.generators_folder)
+        envvars = env.vars(self)
+        envvars.save_script("pkg_config")
```
removed as not necessary

jwillikers reviewed (changes requested):
I think the library name just needs changed. Getting the target name right for the old CMake generators can be tricky with only one component. I'll see if I can find an example.

jwillikers commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,161 @@
+        rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))
+        rmdir(self, os.path.join(self.package_folder, "share"))
+
+    def package_info(self):
+        self.cpp_info.libs = ["qpdf"]
+        self.cpp_info.set_property("cmake_file_name", "qpdf")
+        self.cpp_info.set_property("cmake_target_name", "qpdf::libqpdf")
+        self.cpp_info.set_property("pkg_config_name", "libqpdf")
+
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs.append("m")
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = "qpdf"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "qpdf"
+        self.cpp_info.names["cmake_find_package"] = "qpdf"
+        self.cpp_info.names["cmake_find_package_multi"] = "qpdf"
```
```suggestion
    def self.package_info(self):
        self.cpp_info.set_property("cmake_file_name", "qpdf")

        self.cpp_info.components["libqpdf"].libs = ["qpdf"]
        self.cpp_info.components["libqpdf"].set_property("pkg_config_name", "libqpdf")
        self.cpp_info.components["libqpdf"].set_property("cmake_target_name", "qpdf::libqpdf")
        self.cpp_info.components["libqpdf"].requires.append("zlib::zlib")
        self.cpp_info.components["libqpdf"].requires.append(f"{self.options.with_jpeg}::{self.options.with_jpeg}")

        if self.options.with_crypto == "openssl":
            self.cpp_info.components["libqpdf"].requires.append("openssl::openssl")

        if self.settings.os in ["Linux", "FreeBSD"]:
            self.cpp_info.components["libqpdf"].system_libs.append("m")

        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
        self.cpp_info.filenames["cmake_find_package"] = "qpdf"
        self.cpp_info.filenames["cmake_find_package_multi"] = "qpdf"
        self.cpp_info.names["cmake_find_package"] = "qpdf"
        self.cpp_info.names["cmake_find_package_multi"] = "qpdf"
        self.cpp_info.components["libqpdf"].names["cmake_find_package"] = "libqpdf"
        self.cpp_info.components["libqpdf"].names["cmake_find_package_multi"] = "libqpdf"
```

jwillikers replied:
And of course I typoed after I copied what I tested locally.... sorry about that.

jwillikers commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,161 @@
+from conan.tools.files import patch, export_conandata_patches, get, copy, rmdir
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.microsoft import is_msvc, check_min_vs
+from conan.tools.gnu import PkgConfigDeps
+from conan.tools.env import VirtualBuildEnv, Environment
```
You can remove the unused imports.

```suggestion
from conan.tools.env import VirtualBuildEnv
```

jwillikers reviewed (changes requested):
Okay, I think I figured out the components stuff necessary to get the V1 `cmake_find_package*` working with the correct library name.

cguentherTUChemnitz commented:
> Okay, I think I figured out the components stuff necessary to get the V1 `cmake_find_package*` working with the correct library name.

Thanks for the cleanup here ;) This now looks a lot more mature, than my modified template stub i had for the package_info.

jwillikers commented:
> > Okay, I think I figured out the components stuff necessary to get the V1 `cmake_find_package*` working with the correct library name.
> 
> Thanks for the cleanup here ;) This now looks a lot more mature, than my modified template stub i had for the package_info.

Ha! I think adding the components just for the V1 stuff makes it a whole lot more convoluted. What you had before was great. The problem is it isn't easy to get the V1 `cmake_find_package*` generators to use target names with different names on each side of the `::`. This is easy for V2 where the fully qualified name is provided. For `cmake_find_package*`, it requires using components to get something that looks like `qpd::libqpdf` as a CMake target name.

cguentherTUChemnitz commented:
@uilianries @jwillikers @SSE4 @memsharded   It seems that the build for mac and windows is kind of stuck or doesn't get triggered by the ci system. I see the same behavior here also in my other PR: https://github.com/conan-io/conan-center-index/pull/13356

When i look at the Jenkins link, it shows me an 403 forbidden. So i can't search myself for reasoning here. How to go forward here?

uilianries commented:
@cguentherTUChemnitz Indeed a Mac machine got stuck during all weekend, when trying to build gprc. We canceled that build and it seems be it's working again, but now, the work queue accumulated and will take some time. 

Thank you for pointing that error!

About the CI log access, users are only allowed to read those reported here in GH, the Jenkins is private.

jwillikers commented:
@cguentherTUChemnitz Looks like you just need to patch out the errors from the checks that are failing to now find the `pkg-config` files:

https://github.com/qpdf/qpdf/blob/11.1/libqpdf/CMakeLists.txt#L165-L166
https://github.com/qpdf/qpdf/blob/11.1/libqpdf/CMakeLists.txt#L186-L187

prince-chrismc commented on recipes/qpdf/all/conandata.yml:
```diff
@@ -0,0 +1,11 @@
+sources:
+  "11.1.1":
+    url: [
+      "https://github.com/qpdf/qpdf/archive/refs/tags/v11.1.1.tar.gz",
+    ]
+    sha256: "785edab622a1bc7e25e1537ad2c325005d48c5c7957f7abedff405deb80fa59a"
```
Let's skip on the mirror syntax

```suggestion
    url: "https://github.com/qpdf/qpdf/archive/refs/tags/v11.1.1.tar.gz"
    sha256: "785edab622a1bc7e25e1537ad2c325005d48c5c7957f7abedff405deb80fa59a"
```

prince-chrismc commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,168 @@
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.files import get, copy, rmdir
```
duplicate

```suggestion
```

prince-chrismc commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,168 @@
+            tc.variables["REQUIRE_CRYPTO_GNUTLS"] = False
+            tc.variables["REQUIRE_CRYPTO_OPENSSL"] = True
+        tc.generate()
+        # TODO: after https://github.com/conan-io/conan/issues/11962 is solved
```
https://github.com/conan-io/conan/issues/11962#issuecomment-1273843499

cguentherTUChemnitz replied:
Thanks for pin-point and clarifying at the remote discussion

prince-chrismc commented on recipes/qpdf/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,24 @@
+
+    def test(self):
+        if can_run(self):
+            self.run(os.path.join(self.cpp.build.bindirs[0], "test_package") + " out_empty.pdf", env="conanrun")
```
Could we hardcode the output file name in the exe? since way it's a touch more simple here to replace with new versions

prince-chrismc commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,168 @@
+    def _patch_sources(self):
+        # patch 0001 and 0002 are uniquely appliable, based ony dependency config
+        patches_to_apply = ["0003-exclude-unnecessary-cmake-subprojects.patch"]
+        if self.options.with_crypto == "openssl":
+            patches_to_apply.append("0002-libqpdf-cmake-deps-jpeg-zlib-openssl.patch")
```
Would it be better to always apply the other patch, and than conditionally add in openssl?

```py
apply_conandata_patches(self)
if self.options.with_crypto == "openssl":
    replace_in_files(self, source_folder + "cmakelists", 
                                "PUBLIC JPEG::JPEG ZLIB::ZLIB", "PUBLIC JPEG::JPEG ZLIB::ZLIB OpenSSL::SSL")
```

🤔 not sure it's better but just tossing ideas

cguentherTUChemnitz replied:
Yeah i think your code here is not complete, since the found_package needs to be patched also, but i really like the idea. This would avoid the code-duplication the patch-files. I will update it accordingly

prince-chrismc commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -0,0 +1,168 @@
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "with_crypto": ["native", "openssl", "gnutls"],
```
never head of this project before but what is "native" here? there's no requirements added

cguentherTUChemnitz replied:
It is just going for specific cmake options, telling qpdf to use its own crypto internal provider. Is there some existing working for such a selection? I though about native or internal. I just didn't wont to go for False here, since crypto is still in place, but directly shipped with the qpdf project itself.

cguentherTUChemnitz replied:
I will go for "internal" instead. I think it is slightly more descriptive.

prince-chrismc commented on recipes/qpdf/all/patches/0002-libqpdf-cmake-deps-jpeg-zlib-openssl.patch:
```diff
@@ -0,0 +1,107 @@
++find_package(JPEG REQUIRED)
++find_package(ZLIB REQUIRED)
++find_package(OpenSSL REQUIRED)
++target_link_libraries(${OBJECT_LIB} PUBLIC JPEG::JPEG ZLIB::ZLIB OpenSSL::SSL)
```
I am a little surprised it's not using `OpenSSL::Crypto` given the name of the option

cguentherTUChemnitz replied:
I was going for crypto as wording, since i picked it up from the qpdf dependency docu:
https://qpdf.readthedocs.io/en/stable/installation.html#basic-dependencies

Nevertheless, when there is a more meaningful crypto dependency management naming scheme in other packages, it would be better to sync them (e.g like with_jpeg in severral packages.) When i search through other packages, i think with_ssl is way more often used than with_crypto. So i will change it accordingly.

prince-chrismc reviewed (commented):
super tiny nits and a handful of questions but overall it looks super good

uilianries commented on recipes/qpdf/all/conandata.yml:
```diff
@@ -0,0 +1,8 @@
+    sha256: "785edab622a1bc7e25e1537ad2c325005d48c5c7957f7abedff405deb80fa59a"
+patches:
+  "11.1.1":
+    - patch_file: "patches/0001-libqpdf-cmake-deps-jpeg-zlib.patch"
+    - patch_file: "patches/0002-exclude-unnecessary-cmake-subprojects.patch"
```
Please, add `patch_description` and `patch_type: "conan"` to make easier to understand those patches. Docs: https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields

uilianries reviewed (commented):
Looking good, only a small detail.

prince-chrismc reviewed (dismissed):
LGTM 🚀

prince-chrismc commented:
```
Performing Test WMAIN_WORKED1 - Success
-- Found PkgConfig: C:/J/w/prod/BuildSingleReference@4/.conan/data/pkgconf/1.9.3/_/_/package/0a420ff5c47119e668867cdb51baff0eca1fdb68/bin/pkgconf.exe (found version "1.9.3") 
-- Checking for module 'zlib'
--   Package 'zlib', required by 'virtual:world', not found
-- Checking for module 'libjpeg'
--   Package 'libjpeg', required by 'virtual:world', not found
-- Checking for module 'openssl>=1.1.0'
--   Package 'openssl', required by 'virtual:world', not found
-- 
-- *** Crypto Summary ***
--   GNU TLS crypto enabled: OFF
--   OpenSSL crypto enabled: OFF
--   Native crypto enabled: OFF
--   Default crypto: 
--
```

cguentherTUChemnitz commented:
Sry guys for all the patch-rounds here, but i do not have a windows platform at hand, so  i have to do quite some ping-pong with the ci.

cguentherTUChemnitz commented:
@uilianries What is here to do, when crucial dependency like the default used openssl is missing here on ci as prebuilt package?

cguentherTUChemnitz commented:
@uilianries @prince-chrismc I think we are finally here in a mergeable state with openssl conan support for all major platforms ;)

prince-chrismc commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -62,8 +62,9 @@ def layout(self):
-        self.requires("zlib/1.2.12")
+        self.requires("zlib/1.2.13")
         if self.options.with_ssl == "openssl":
+            # TODO: update to 3.x branch of openssl, when zlib is able to handle it
```
Isn't the dependencies the other way around? 😖

cguentherTUChemnitz replied:
A yes, your are right. I think it it not necessary to add the comment here after all, since updating dependencies is like an on-going todo... I would just now remove the comment to solve the discussion here.

prince-chrismc commented on recipes/qpdf/all/conanfile.py:
```diff
@@ -119,12 +124,24 @@ def generate(self):
         apply_conandata_patches(self)
+        # we generally expect to have one crypto in-place, but need to patch the found mechanics
+        # since we avoid currently the correct pkg_config
+        replace_in_file(self, os.path.join(self.source_folder, "libqpdf", "CMakeLists.txt"),
```
This can go in the parch I think

cguentherTUChemnitz replied:
Mh, it is possible, but i would like to have the handling for the FOUND_CRYPTO, USE_CRYPTO_OPENSSL, USE_CRYPTO_GNUTLS, USE_CRYPTO_NATIVE next to each other, since they are already contextwise near to each other.
